### PR TITLE
Security hardening + heartbeat backend + live-device test results

### PR DESCRIPTION
**Pull request by dallanwagz**
## Summary

Combined PR superseding both #23 (security hardening) and #24 (heartbeat
backend), and adding **18 captured live heartbeats** from the real clock
device as the new top commit.

The full stack in one PR:

| Commit | What |
| --- | --- |
| `a4b4910` | Add REST API for automated testing and device management |
| `725ecdd` | Security hardening: fix 13 of 16 audit findings |
| `4b17ab8` | Add security hardening test results |
| `8a2b7ea` | Fix markdownlint violations in security docs |
| `a755d8b` | Extract security helpers into testable module; convert to pytest |
| `f27d910` | Add OTA rollback test results with full crash-loop coverage |
| `6f728cf` | Remove tool attribution from test result docs |
| `fd289ab` | Add device heartbeat telemetry to calendar fetch |
| `3884f35` | Add WagFam backend server with heartbeat tracking and calendar API |
| `70f63a5` | Add E2E test results for heartbeat server (33/33 pass) |
| `a73b6bc` | **Add live end-to-end test results for heartbeat telemetry** ← new |

The only commit not previously reviewed in #23 or #24 is the last one
(`a73b6bc`), which adds `docs/TEST_RESULTS_HEARTBEAT_LIVE.md`.

## What the live-test commit adds

Captures **18 real heartbeats** from the live clock (chip_id `5fc8ad`,
IP `192.168.168.66`) over a 16-minute window with the refresh interval
temporarily dropped to 1 minute. Closes the gap left by PR #24's
`TEST_RESULTS_HEARTBEAT_SERVER.md` which was simulation-only.

- Live device → local TLS server (FastAPI on `https://10.10.2.178:8443`)
  via inter-subnet routing; self-signed cert accepted by the firmware's
  `BearSSL::WiFiClientSecure::setInsecure()` client
- OTA round-trip to flash a build of this branch — needed because the
  device's prior 3.08.0-wagfam binary predated the heartbeat commit and
  was sending no `chip_id` query string
- 17 nominal cycles within 60 ± 3 s at `minutes_between_data_refresh = 1`
- Heap stable across 18 fetches (33,192–34,344 B, σ ≈ 250 B — no leak)
- Server upsert: one row per `chip_id`, latest values reflected, real LAN
  IP (`192.168.168.66`)

## Restored after live test

| Setting | Original | Test | Restored? |
| --- | --- | --- | --- |
| `wagfam_data_url` | GitHub raw URL | local TLS endpoint | yes |
| `wagfam_api_key` | (40-char GH PAT) | `livetest-key-001` | yes |
| `minutes_between_data_refresh` | 60 | 1 | yes |
| `OTA_SAFE_URL` | dead `lhr.life` tunnel (HTTP 503) | local test host | no — `/conf.txt` is a protected path on `/api/fs/write`; both pre and post values are unreachable, operational state unchanged |

## Diagnostic finding

The device reported `version: "3.08.0-wagfam"` but its binary predated
the heartbeat commit. **Recommendation:** bump `VERSION` when wire-visible
behavior changes, otherwise field debugging is ambiguous.

## Supersedes

- #23 — security hardening (all commits included)
- #24 — heartbeat backend + server (all commits included)

#23 and #24 are left open for now; close them after merging this.

## Test plan

- [x] `make lint-markdown` — 0 errors / 12 files
- [x] All 18 captured heartbeats had every telemetry field populated, HTTP 200
- [x] Cadence within 60 ± 3 s for 17 nominal cycles
- [x] Heap stable, no leak signal
- [x] Server upsert produces one stable row, latest values reflected
- [x] Clock config fully restored (verified via post-test `GET /api/config`)
- [x] All security hardening + heartbeat-server tests already validated in #23/#24

### CODE REVIEW

File: CLAUDE.md [hunk 1]
```diff
@@ -166,6 +168,19 @@ Expected JSON format:
...
+### Device Heartbeat
+
+Each calendar fetch includes device telemetry as query parameters:
+
+```text
+GET /data_source.json?chip_id=5fc8ad&version=3.08.0-wagfam&uptime=1234567&heap=32496&rssi=-62
+```
+
+This lets a backend identify and monitor all deployed clocks without any additional
+connections. Static JSON hosts (e.g., GitHub raw) ignore the query params gracefully.
```
Comment (jrwagz) [inline CLAUDE.md]: I need you to please provide evidence and proof of the fact that GitHub raw will ignore these new query params gracefully.  If we roll this out, and that claim isn't true, all deployed clocks will immediately croak.

Furthermore, update your CLAUDE.md so that in the future, no such claims are made without proving evidence of that fact, in the form of either automated tests, or links to official documentation.

File: tests/native/test_security_helpers/test_security_helpers.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    TEST_ASSERT_EQUAL_STRING("example.com",
+        extractDomain("https://example.com").c_str());
+}
+
+void test_extract_domain_empty_string() {
+    TEST_ASSERT_EQUAL_STRING("", extractDomain("").c_str());
+}
+
+void test_extract_domain_no_scheme() {
+    TEST_ASSERT_EQUAL_STRING("", extractDomain("example.com/path").c_str());
```
Comment (jrwagz) [inline tests/native/test_security_helpers/test_security_helpers.cpp]: please improve extractDomain so that it can indeed extract `example.com` as the domain when provided `example.com/path`.  While this isn't a testcase we would hit on our clocks, it seems like a perfectly valid use of the function that could easily be supported.

File: marquee/SecurityHelpers.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  return url.substring(start, end);
+}
+
+// SEC-03: Validate firmware URL domain against the calendar data source domain
+bool isTrustedFirmwareDomain(const String &firmwareUrl, const String &calendarUrl) {
+  String fwDomain = extractDomain(firmwareUrl);
+  String calDomain = extractDomain(calendarUrl);
+  if (fwDomain == "" || calDomain == "") return false;
+  return fwDomain == calDomain;
+}
```
Comment (jrwagz) [inline marquee/SecurityHelpers.cpp]: I'm going to challenge the assertion that we will always be serving the firmware and calendars from the same domain.  I don't think this will absolutely hold true in production, and therefore we need a way to have a few static and trusted domains complied into the binary that should be trusted for firmware urls no matter what the calendar URL is.

File: server/app/routers/devices.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+@router.get("/api/v1/devices/{chip_id}", response_model=DeviceResponse)
+def get_device(chip_id: str, db: Session = Depends(get_db)):
+    device = db.query(Device).filter(Device.chip_id == chip_id).first()
+    if not device:
+        raise HTTPException(status_code=404, detail="Device not found")
+    return device
+
+
+@router.patch("/api/v1/devices/{chip_id}", response_model=DeviceResponse)
+def update_device(chip_id: str, body: DeviceUpdate, db: Session = Depends(get_db)):
```
Comment (jrwagz) [inline server/app/routers/devices.py]: Please update the documentation for this endpoint so that it's more clear that we are only updating the device's name.  The way the endpoint reads it seems like the entirety of the device could be modified, but in fact only the name can be adjusted.  Change the function name to something like update_device_name and the api endpoint to be a POST to /api/v1/devices/{chip_id}/update_name so that it's much more clear what is going on.

File: server/tests/test_calendar_event.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
+"""Tests for CalendarEvent, adapted from jrwagz/wagfam-clocks-data-source.
```
Comment (jrwagz) [inline server/tests/test_calendar_event.py]: You say that these tests and the source code are "adapted" from jrwagz/wagfam-clocks-data-source.  Please provide exact details on what exactly has changed from the source, and what tests have been written to prove the changes are correct.  If no changes to the source or tests have completed, please provide evidence as such by the comparison of the files at the specific git commit hash on the source repo.

File: server/app/services/calendar_gen.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
+"""Calendar message generation adapted from jrwagz/wagfam-clocks-data-source."""
```
Comment (jrwagz) [inline server/app/services/calendar_gen.py]: If this code was "adapted" from the source, please provide evidence and documentation of exactly what changed so that those changes can be reviewed.  If no changes were made, then be very clear about that and show exactly what version of the code was copied.  If you don't need to maintain your own changes to the files from that repo, then add that repo as a submodule so that it's 100% clear that no changes are being made, and the code is just leveraged as-is.

**Review by jrwagz — CHANGES_REQUESTED** [review]

File: CLAUDE.md (hunk 1 above)
Comment (dallanwagz) [reply CLAUDE.md]: Addressed in bca26e6:

**Evidence (positive):** Both `?chip_id=...&version=...&uptime=...&heap=...&rssi=...` and a no-param fetch against `https://raw.githubusercontent.com/jrwagz/marquee-scroller/master/README.md` return the **same** sha256 (`bb50d78f1d1c2e4a3b5166ece4f4907eb94e4f2b1475a3c6173d70269e8399cc`), HTTP 200, and content-length 6377. Each of the 5 individual params verified the same way.

**Evidence (negative — important caveat I uncovered):** `?token=...` IS interpreted by raw.githubusercontent.com as an auth attempt — sending a fake token against a private repo returns HTTP 404 instead of the file content. So `token` must never become a heartbeat parameter name in the future.

**Pinned as a regression test:** `server/tests/test_github_raw_compat.py` (8 tests) makes the real HTTPS request and asserts both directions. Skipped when offline. CI will catch any GitHub behavior change.

**CLAUDE.md policy:** Added a "Back assertions about external services with evidence" rule under Development Practices that points at this exact test as the pattern to follow. Future external-service claims must come with either an automated test that hits the real service, or a citation to official documentation.

File: tests/native/test_security_helpers/test_security_helpers.cpp (hunk 2 above)
Comment (dallanwagz) [reply tests/native/test_security_helpers/test_security_helpers.cpp]: Addressed in bca26e6.

`extractDomain` now handles scheme-less URLs and a few other edge cases that the original implementation got wrong:

| Input | Before | After |
| --- | --- | --- |
| `example.com/path` | `""` | `example.com` |
| `example.com` | `""` | `example.com` |
| `example.com:8080/file` | `""` | `example.com` |
| `https://user:pass@example.com/path` | `user:pass@example.com` | `example.com` (userinfo stripped) |
| `example.com?foo=bar` | `""` | `example.com` (query stripped) |
| `example.com#section` | `""` | `example.com` (fragment stripped) |

The negative test you flagged (`test_extract_domain_no_scheme`) is replaced with three positive cases — `_no_scheme_with_path`, `_no_scheme_bare_host`, `_no_scheme_with_port` — plus three additional edge cases for userinfo, query, and fragment stripping. All previously passing tests still pass; total native tests went from 84 to 104.

File: marquee/SecurityHelpers.cpp (hunk 3 above)
Comment (dallanwagz) [reply marquee/SecurityHelpers.cpp]: Addressed in bca26e6. Agreed — the same-domain assumption is too restrictive for production.

**New design:**

`marquee/Settings.h` now declares a compile-time allowlist macro:

```c
#ifndef WAGFAM_TRUSTED_FIRMWARE_DOMAINS
#define WAGFAM_TRUSTED_FIRMWARE_DOMAINS ""
#endif
```

Override at build time:

```bash
pio run -e default --build-flag '-DWAGFAM_TRUSTED_FIRMWARE_DOMAINS="\"cdn.example.com,releases.example.com\""'
```

`isTrustedFirmwareDomain` now takes a third `trustedAllowlist` parameter and returns true if the firmware URL's domain matches **either**:

1. an entry in the compile-time allowlist (new), OR
2. the calendar URL's domain (existing behavior, preserved)

The runtime call site at `marquee.ino:855` passes `WAGFAM_TRUSTED_FIRMWARE_DOMAINS`. Default value is empty, so behavior is unchanged for anyone not setting the macro at build time — backwards-compatible with the existing deployment.

**New tests** (15 added, all passing):

- `isInTrustedDomainList` covers single entry, multi-entry positions, whitespace trimming, empty/null inputs, and substring/prefix-match rejection.
- `isTrustedFirmwareDomain` with allowlist covers: allowlist overrides calendar mismatch, unlisted domain rejected, fallback-to-calendar still works, empty allowlist behaves identically to the old contract.

I left the macro empty by default rather than picking a specific domain set — happy to populate with whatever production CDN domain you want as a follow-up commit if you have one in mind.

File: server/app/routers/devices.py (hunk 4 above)
Comment (dallanwagz) [reply server/app/routers/devices.py]: Addressed in bca26e6.

**API surface change:**

| Before | After |
| --- | --- |
| `PATCH /api/v1/devices/{chip_id}` body `{"name": "..."}` | `POST /api/v1/devices/{chip_id}/update_name` body `{"name": "..."}` |
| Function: `update_device` | Function: `update_device_name` |
| Schema: `DeviceUpdate` (`name: str \| None`, optional) | Schema: `DeviceNameUpdate` (`name: str`, **required**) |
| No endpoint description | Explicit `summary` + `description` in the route decorator stating "Updates ONLY the device's `name` field — no other fields can be modified through this endpoint" |

**Tests added** (and the old PATCH path proven gone):

- `test_update_device_name` — happy path
- `test_update_device_name_requires_auth` — 401 without token
- `test_update_device_name_404_when_unknown_device` — 404 on bad chip_id
- `test_update_device_name_rejects_missing_name` — 422 when body lacks `name` (Pydantic enforced)
- `test_update_device_name_ignores_other_fields` — extra fields like `version`/`uptime_ms` in the body are silently dropped by Pydantic and **do not mutate** the device's telemetry
- `test_old_patch_endpoint_returns_method_not_allowed` — regression test that PATCH on the old path returns 405

**Docs updated:** `docs/TEST_RESULTS_HEARTBEAT_SERVER.md` Test 5 now shows the new POST shape and notes the rename was in response to this review feedback.

File: server/tests/test_calendar_event.py (hunk 5 above)
Comment (dallanwagz) [reply server/tests/test_calendar_event.py]: Addressed in bca26e6.

**Source pinned:** `jrwagz/wagfam-clocks-data-source@ddd0f0509f74dc01f73a764c817b561e2f864afc` (`main` HEAD as of 2026-04-26).

**What changed in this file (`tests/test_calendar_event.py` → `server/tests/test_calendar_event.py`):**

- Replaces the source's autouse `monkeypatch` fixture for `todays_date()` with explicit `today=TODAY` passed to the new `CalendarEvent(today=...)` constructor parameter (see comment on `calendar_event.py` for *why* that parameter was added).
- Adds a `make_event(data)` helper that closes over `TODAY` so each test reads cleanly.
- Drops verbose docstrings on test functions (the test name itself documents intent).
- Asserts `next_occurrence == datetime.date(...)` instead of separate `.year`/`.month`/`.day` checks (equivalent, more concise).

**Test coverage is preserved 1:1** — every test in the source file has a corresponding test in the adapted file asserting the same behavior.

**Evidence checked in:**

- `docs/CALENDAR_ADAPTATION.md` — narrative + section per file
- `docs/adaptations/test_calendar_event.diff` — full unified diff against source

**Verification recipe** (also in CALENDAR_ADAPTATION.md): clone the source repo at the pinned commit, run `diff -u` between source and adapted, then `diff -u` that against the checked-in patch. Output should be empty. I ran it three times before pushing — all pass.

File: server/app/services/calendar_gen.py (hunk 6 above)
Comment (dallanwagz) [reply server/app/services/calendar_gen.py]: Addressed in bca26e6.

**Source pinned:** `jrwagz/wagfam-clocks-data-source@ddd0f0509f74dc01f73a764c817b561e2f864afc`.

**This file's source is actually `generate_data_source.py`** — a CLI script, not a library module. The adaptation is fundamental: drop the `argparse` + `main()` plumbing, and return values from `generate_messages()` instead of writing JSON to disk. `load_calendar_data()` is a small new server-side wrapper for instantiating events from disk.

**Behavioral algorithm is unchanged** — the sort/walk/cap-at-N logic is the same as the source.

**On the submodule suggestion:** I went down that path and ended up *not* recommending it. Reasons documented in `docs/CALENDAR_ADAPTATION.md` under "Why not a submodule":

1. The source is shaped as a CLI script, not a library — its top-level Python files `import argparse` and run `main()`. There's no clean importable public API.
2. No published version, no `pyproject.toml` package metadata, no `__init__.py`. A submodule pin would still require `sys.path` manipulation to use.
3. The `today=` parameter divergence on `CalendarEvent` is intentional and we depend on it for testability. With a submodule, every upstream change would re-require manual re-application of that adaptation.

The doc proposes the right long-term fix: refactor the source repo upstream into a real `calendar` subpackage with `__init__.py` and `pyproject.toml`, publish wheels (or make it git-installable), then drop the adapted copies here. Out of scope for this PR but happy to file an issue on `wagfam-clocks-data-source` if you want to track it.

**Evidence checked in:**

- `docs/CALENDAR_ADAPTATION.md` — narrative + per-file change details + verification recipe
- `docs/adaptations/calendar_gen.diff` — full unified diff against `generate_data_source.py`
- File header in `calendar_gen.py` now points at the doc and the diff

**Verification recipe** runs `diff -u` between source-at-pinned-commit and adapted, then `diff -u` against the checked-in patch — output empty when in sync. Verified all three files (calendar_event, calendar_gen, test_calendar_event) before push.

File: server/Makefile [hunk 7]
```diff
@@ -0,0 +1,22 @@
+.PHONY: run test docker-build docker-up docker-down install
+
+install:
+	pip install -e ".[dev]"
+
+run:
+	uvicorn app.main:app --reload
+
+test:
+	python -m pytest tests/ -v
```
Comment (jrwagz) [inline server/Makefile:10]: Please integrate these server tests into the top level makefiles `make test` target.  Specifically there should be a `test-server` targer which the `test` target depends on, and the `test-server` target should execute the tests in the server folder.

Furthermore, the tests in this folder need to be containerized to ensure that the versions of tools used in the testing are consistent across developers.  For example, I was unable to get these tests to run locally, here is the errors I got:

```
$ cd server 
(marquee-scroller) 
 16:16:04 projects/marquee-scroller/server 
$ make test
python -m pytest tests/ -v
ImportError while loading conftest '/Users/justinwagner/projects/marquee-scroller/server/tests/conftest.py'.
tests/conftest.py:6: in <module>
    from sqlalchemy import create_engine
E   ModuleNotFoundError: No module named 'sqlalchemy'
make: *** [test] Error 4
(marquee-scroller) 
 16:16:05 projects/marquee-scroller/server 
$ make install
pip install -e ".[dev]"
Obtaining file:///Users/justinwagner/projects/marquee-scroller/server
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
INFO: pip is looking at multiple versions of wagfam-server to determine which version is compatible with other requirements. This could take a while.
ERROR: Package 'wagfam-server' requires a different Python: 3.9.6 not in '>=3.10'
make: *** [install] Error 1
(marquee-scroller) 
 16:16:42 projects/marquee-scroller/server 
$ which pip
/Users/justinwagner/projects/marquee-scroller/.venv/bin/pip
(marquee-scroller) 
 16:17:24 projects/marquee-scroller/server 
$ pip --version
pip 26.0.1 from /Users/justinwagner/projects/marquee-scroller/.venv/lib/python3.9/site-packages/pip (python 3.9)
(marquee-scroller) 
 16:17:29 projects/marquee-scroller/server 
$ pip3 --version
pip 26.0.1 from /Users/justinwagner/projects/marquee-scroller/.venv/lib/python3.9/site-packages/pip (python 3.9)
(marquee-scroller) 
 16:17:32 projects/marquee-scroller/server 
$ python3 --version
Python 3.9.6
(marquee-scroller) 
 16:17:37 projects/marquee-scroller/server 
$ python3.12 --version
Python 3.12.3
```

**Review by jrwagz — CHANGES_REQUESTED** [review]
This branch has merge conflicts which need to be fixed, and as a result, CI tests are not currently running on it.  Also, see my comments regarding making server tests more reliable.

File: server/Makefile (hunk 7 above)
Comment (dallanwagz) [reply server/Makefile:10]: Done in 455a750.

- Added `test-server` target to the root `makefile` that builds the server Docker image and runs `pytest` inside it — no local Python 3.10+ needed
- `test` now depends on both `test-native` and `test-server`
- Server `Dockerfile` installs `.[dev]` so pytest + httpx are available in the container
- Fixed `test_github_raw_compat.py` to use `httpx` (already a dev dep) instead of `requests` (wasn't installed)

All 41 tests pass in the container:
```
make test-server
```

**Review by jrwagz — CHANGES_REQUESTED** [review]
These changes still cannot be merged because they have merge conflicts on marquee.ino, please solve those merge conflicts and update the PR accordingly.

File: makefile [hunk 8]
```diff
@@ -106,8 +108,19 @@ test-scripts: .passwd .pytest-image
...
+SERVER_IMAGE:=wagfam-server-test
+
+.PHONY: test-server
+test-server:
+	docker build -t $(SERVER_IMAGE) server/
+	docker run \
+		--rm $(DOCKER_TTY_ARGS) \
+		-e WAGFAM_WAGFAM_API_KEY=test-key \
+		$(SERVER_IMAGE) \
+		python -m pytest tests/ -v
```
Comment (jrwagz) [inline makefile]: Please improve test-server so that it generates the junit formatted test results (just like test-native and test-scripts) and then modify the CI workflow to include those tests results just like is done for test-native and test-scripts.

**Review by jrwagz — CHANGES_REQUESTED** [review]

File: makefile (hunk 8 above)
Comment (dallanwagz) [reply makefile]: Done in 27c0552.

- `test-server` now mounts `artifacts/` into the container and writes JUnit XML to `artifacts/test-server-results.xml`
- Added "Server Test Report" step in CI workflow using `dorny/test-reporter@v3`, matching the pattern of Native and Scripts reporters

**Review by jrwagz — APPROVED** [review]
LGTM
